OpenSim/Simulation/Model/Model.cpp

```diff
@@ -1875,6 +1851,50 @@ bool Model::getAllControllersEnabled() const{
 void Model::setAllControllersEnabled( bool enabled ) {
     _allControllersEnabled = enabled;
 }
+bool Model::isControlsCacheValid(const SimTK::State& state) const{
+    return getDefaultSubsystem().isCacheValueRealized(state,_modelControlsIndex);
+}
+//------------------------------------------------------------------------------
+//            OVERRIDDEN METHOD TO INITIALIZE CONTROLS CACHE
+//------------------------------------------------------------------------------
+void Model::extendRealizeTopology(SimTK::State& s) const
+{    
+    Super::extendRealizeTopology(s);
+
+    Model *mutableThis = const_cast<Model *>(this);
+
+    // Create the shared cache that will hold all model controls
+    // This must be created before Actuator.extendAddToSystem() since Actuator will append 
+    // its "slots" and retain its index by accessing this cached Vector
+    // value depends on velocity and invalidates dynamics BUT should not trigger
+    // recomputation of the controls which are necessary for dynamics
+    mutableThis->_modelControlsIndex = getDefaultSubsystem().allocateCacheEntry(
+        s, Stage::Velocity, Stage::Acceleration, new Value<Vector>());
+}
```

Comment:
Using the Component::addCacheVariable should be virtually identical to this except that it uses a lazy and in this case and in the previous version the controls must be evaluated by Acceleration.

The reason that this is now thread-safe is because you force the evaluation in the extendRealizeVelocity() which is probably the right thing to do. And now it is no longer lazily evaluated either.